### ISSUE TITLE
Multiplayer skipping turns on load / AI playing twice in one turn

### ISSUE THREAD

**Issue by raphaell666** [LoneGazebo/Community-Patch-DLL#4166]
This is related to this issue: https://github.com/LoneGazebo/Community-Patch-DLL/issues/2576, and this pull request: https://github.com/LoneGazebo/Community-Patch-DLL/pull/4002. I am still looking into this, and would appreciate any help as this issue is very relevant in my multiplayer games (and I believe for others too).

I've been told by G that this issue is possibly not present when saving two times one after the other (I have not tested this yet, but plan on doing so soon). However, I do not believe that this is enough of a fix for this issue, due to the constant desyncs and associated problems in multiplayer. On long games you are frequently forced to load previous autosaves due to desync bugs, and when loading autosaves there is no way to "double save" to avoid this issue.

The AutoSave function from the "structure" pointed to by gDLL apparently has the options `AutoSave(bool bInitial = false, bool bPostTurn = false)` as seen in ICvDLLUtility.h. However, it seems the only things these bools change are the name of the AutoSaves (impressively enough). I may be wrong because I haven't extensively tested it, but `bInitial` changes the autosave name to "AutoSave_Initial" while `bPostTurn` changes the autosave name to "AutoSave_Post_xxx", and both don't seem to change anything regarding the actual save itself. I could not find where this structure (or whatever it is) pointed to by this gDLL is defined, does anyone know any details about it? I really couldn't find anything about it.

This is the original civ5 code (inside CvGame.cpp `updateMoves()`) that I will mention below:
```
if(activatePlayers)
{
	if (isOption(GAMEOPTION_DYNAMIC_TURNS) || isOption(GAMEOPTION_SIMULTANEOUS_TURNS))
	{//Activate human players who are playing simultaneous turns now that we've finished moves for the AI.
		// KWG: This code should go into CheckPlayerTurnDeactivate
		for(iI = 0; iI < MAX_PLAYERS; iI++)
		{
			CvPlayer& player = GET_PLAYER((PlayerTypes)iI);
			if(!player.isTurnActive() && player.isHuman() && player.isAlive() && player.isSimultaneousTurns())
			{
				player.setTurnActive(true); 
			}
		}
	}
}
```

I've been testing and playing around with the fix from @davenye (https://github.com/LoneGazebo/Community-Patch-DLL/pull/4002) with my multiplayer games, and I noticed a few things/have a few thoughts that might be relevant:

- The issue seems to be exactly during the call to `player.setTurnActive(true)`. If I move the fix to before it, the fix has no effect. If I move to after it, it works;
- In other situations in the code, when `gDLL->sendTurnUnready()` is used,  `setEndTurn` is also called, I am not exactly sure why but I am testing the fix using it as well (you can see what I am doing here: https://github.com/raphaell666/Community-Patch-DLL);
- `setEndTurn` and `setTurnActive` seem to do different things, but `setEndTurn` does not solve this problem alone;
- `gDLL->sendTurnUnready()` only works for the local/active player, therefore players offline when loading the game will still lose a turn (e.g. pitboss games). I could not find a way around this;
- The function `CancelActivePlayerEndTurn()` looks perfect to solve this issue as it also uses `gDLL->sendTurnUnready()`, but it in fact does not help with this issue, I think because `player.isEndTurn()` is False, only `gDLL->HasReceivedTurnComplete(player.GetID())` seems to return True and catch this issue. This makes me wonder if was there a wrong `gDLL->sendTurnComplete()` somewhere? Why is this issue only noticed and fixed by using the functions pointed to by gDLL?

TL;DR: the fix from @davenye seems to work in most cases (I will be using it in my games from now on), but I wonder if we could find the source of this issue and a more elegant way to fix this once and for all, preferably not requiring that all players be online during the load (though I agree this is still acceptable). My main question would be if there is anyway to read the source of *gDLL, or if anyone knows more about it?

**Comment by LoneGazebo** [comment]
gDLL is either the graphical DLL or the netcode depending on where called. We can’t access it.

**Comment by raphaell666** [comment]
That's a shame, but thank you for the information.

I have been doing some more extensive tests, and indeed the different AutoSave options are useless and all file contents match exactly between them. The only difference is indeed the name of the file.

I also documented my tests with comments in the source files. I think I got to a pretty OK solution, 99% of the credit to @davenye, with a small addition to bring it in line to how turn canceling is done in other places of the code, you can see it here: https://github.com/raphaell666/Community-Patch-DLL/commit/c0c017144bf8ac3a46b0b93db2e25c093d316caa. I basically added `player.setEndTurn(false)` if `gDLL->sendTurnUnready()` is successful, as is done everywhere else in the code, in hopes that this could help avoid unexpected issues.

I will keep on playing my multiplayer game (now with VP updated to 04/11 and with the fixes mentioned above), and report if I find any issues.

**Comment by LoneGazebo** [comment]
@raphaell666 thanks, it's hard to get MP debug scenarios setup so feedback always appreciated. Let me know if I fixed the deal cancellation bug you noticed. 
The end turn element that you fixed may or may not have an impact on desyncs, we'll see (for example, if the net DLL isn't waiting on barbarians to finish their turn, it could cause a desync).

**Comment by raphaell666** [comment]
I moved the fix to the CvPlayer::SetTurnActive function, as it seems to be closer to the root of the issue (https://github.com/raphaell666/Community-Patch-DLL/commit/42794738301d5de2463f9bdf97e57be34294076f). I hope to play around with the SetTurnActive function now to see at what point exactly it bugs out. In addition, I noticed SetTurnActive(true) already calls setEndTurn(false), so that extra call I added was redundant.

Regarding the MP game with this fix, we played for around 3 hours, got one weird non-repeatable CTD  and successfully reloaded the last AutoSave without the AI double turning on us (phew! we were at war). So far everything seems to be going perfectly in the game. This game-loading fix really makes a huge difference for these relatively short duration sessions. We can now play during wars without being afraid of crashes and can stop playing during wars without being destroyed by the AI  getting double the moves when reloading.

**Comment by raphaell666** [comment]
Update: I have pinpointed the issue to when `m_bTurnActive = bNewValue;` is set within `CvPlayer::setTurnActive`. 

All that I know is that setting `m_bTurnActive` to true causes `CvPlayer::isTurnActive()` to also return true. Therefore some function waiting for `isTurnActive()` to be true could be what is causing this issue. I will see if I can learn anything else, but I feel that I am close to the limit of my paltry debugging skills.

This is the best version of the fix that I have tested so far: https://github.com/raphaell666/Community-Patch-DLL/commit/aa713d376713eb7c32094b5fcc911c56f1a56629
With this positioning of the fix, the player icons don't even stop flashing when loading the MP savegame. Previously they would briefly stop flashing (indicating turn inactive) before lighting up again due to the fix. Without the fix the player icons would simply stop flashing and the AI would get their free second play in one turn.

I will also continue playing multiplayer games with this latest version of the fix, and report if I notice anything.

If someone could review the code that I am writing in these latest versions I would really appreciate it.

**Comment by LoneGazebo** [comment]
Honestly the best test is playtime - if it all works as it should I'll check the code for issues.

G

**Comment by LoneGazebo** [comment]
closing b/c duplicated
